File: hard/trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int lft[size(height)], rt[size(height)];
        int a=0;
        for (int i=0; i<size(height); ++i) {
            lft[i]=a;
            a=max(a,height[i]);
        }
        a=0;
        for (int i=size(height)-1; i>=0; --i) {
            rt[i]=a;
            a=max(a,height[i]);
        }
        int amt,res=0;
        for (int i=0; i<size(height); ++i) {
            amt=min(lft[i],rt[i])-height[i];
            if (amt>0)
                res+=amt;
        }
        return res;
    }
};

File: hard/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        leftmax=[]
        rightmax=[0]*len(height)
        msf=0
        for i in height:
            leftmax.append(msf)
            msf=max(msf,i)
        msf=0
        for i in range(len(height)-1,-1,-1):
            rightmax[i]=msf
            msf=max(msf,height[i])
        res=0
        for i in range(0,len(height),1):
            amt=min(leftmax[i],rightmax[i])-height[i]
            if amt>-1:
                res+=amt
        return res
